### PR TITLE
Record epoch times

### PR DESCRIPTION
**Pull request by CarolinaFurtado**
It works: 
image
image

### CODE REVIEW

**Comment by rak5216** [comment]
simple enough, nice job! one thing we should consider yet is whether we can take the output spreadsheet here and easily append it to the standard metrics output. would be good to have everything in one sheet, and easiest way would be to combine the spreadsheets probably. given that both are csv, i imagine something like this maybe:

https://stackoverflow.com/questions/55639115/combining-csvs-with-different-columns-using-pandas-with-key-column

**Comment by rak5216** [comment]
also, we def now want to plot epoch time too and add to mosaic

**Comment by CarolinaFurtado** [comment]
@rak5216 adding to the mosaic is not straight forward, because it is in a different file and also, does not have train and validation separately. Would having individual plots be enough?

**Comment by rak5216** [comment]
you could just make it the first plot in the mosaic with an if statement in the loop, no? that would be ideal

**Comment by CarolinaFurtado** [comment]
@rak5216 I was able to add it to metrics! super easy, in the end... 
image

now working on the plots should be easy

**Comment by CarolinaFurtado** [comment]
## individual plots (i'll add seconds in the y axis)
image

image


## Mosaic plot
image


## Metrics.csv
image

@rak5216

File: train_segmentation_model.py [hunk 1]
```diff
@@ -35,12 +52,16 @@ def generate_plots(metric_names, epochs, compiled_model, results, plots_dir, num
...
             fig2, axes = plt.subplots(nrows=num_rows, ncols=num_cols, squeeze=False)
 
-        # plot the train and validation curves
-        for split in ['train', 'validate']:
-            key_name = metric_name
-            if split == 'validate':
-                key_name = 'val_' + key_name
-            axes[counter_rows, counter_col].plot(range(epochs), results.history[key_name], label=split)
+        # plot
+        if metric_name in ['epoch_time_in_sec', 'total_elapsed_time_in_sec']:  # plot the total time and epoch time separatly
```
Comment (rak5216) [inline train_segmentation_model.py:56]: typo on separately in comment

File: train_segmentation_model.py [hunk 2]
```diff
@@ -9,20 +9,37 @@
...
 
 metadata_file_name = 'metadata.yaml'
 
 tmp_directory = Path('./tmp')
 
 
-def generate_plots(metric_names, epochs, compiled_model, results, plots_dir, num_rows=1, num_cols=1):
+class timecallback(Callback):
+    def __init__(self):
+        # use this value as reference to calculate cummulative time taken
```
Comment (rak5216) [inline train_segmentation_model.py:29]: typo in cumulative

File: train_segmentation_model.py [hunk 3]
```diff
@@ -9,20 +9,37 @@
...
-from tensorflow.keras.callbacks import ModelCheckpoint, TensorBoard, CSVLogger
+from tensorflow.keras.callbacks import ModelCheckpoint, TensorBoard, CSVLogger, Callback
 from image_utils import TensorBoardImage, ImagesAndMasksGenerator
 import git
 from gcp_utils import copy_folder_locally_if_missing
 from models import generate_compiled_segmentation_model
 from metrics_utils import global_threshold
 from local_utils import local_folder_has_files, getSystemInfo, getLibVersions
+import time
+import csv
```
Comment (rak5216) [inline train_segmentation_model.py:20]: is this still needed?

**Review by rak5216 — COMMENTED** [review]
looks great, just a few typos and maybe unused import. does the results.history list of metrics automatically include the data from the timecallback then (maybe via logs[...])? interesting

File: train_segmentation_model.py (hunk 3 above)
Comment (CarolinaFurtado) [reply train_segmentation_model.py:20]: time is, csv is not, you are right. removed

**Comment by CarolinaFurtado** [comment]
@rak5216  the new callback edits the logs by adding a new entry to the dictionary everytime a epoch ends. The `callbacks` in

```
    results = compiled_model.fit(
        train_generator,
        steps_per_epoch=len(train_generator),
        epochs=epochs,
        validation_data=validation_generator,
        validation_steps=len(validation_generator),
        callbacks=[model_checkpoint_callback, tensorboard_callback, time_callback, csv_logger_callback]
    )
```

are called sequentially. That is why `csv_looger_callback` is called last. This way it already accounts for the new entry created by `time_callback`. Pretty simple in the end... But I only found one link that explained this well.
